### PR TITLE
Add methods for polymorphic associations to decision reviews

### PR DESCRIPTION
**Pull request by yoomlam**
TLDR: This PR is an enabler for adding new tables in the ETL DB for Paul S.'s D.O.C. (Decision Output Calculator = "required decision output of production attorneys"). Instead of writing fragile SQL queries, this PR uses Rails incantations so that future changes to application code are reflected in the ETL transformations.

Explanation: The D.O.C. requires an ETL table that joins `DecisionDocument` records with `DecisionIssue`, along with some data from the associated `Appeal` record. This could be done in an SQL query, but changes to our Rails models would also require changes to the SQL query, so this PR creates Rails associations so that Rails can generate the appropriate SQL. This would be straightforward if it wasn't for polymorphic associations to the `Appeal` model -- defining [`has_many :decision_issues through: appeal`](https://coderwall.com/p/fvjcwq/preloading-scoped-has_many-through-associations) in `DecisionDocument` doesn't work because `appeal` is polymorphic. This PR addresses this limitation by dynamically defining associations to `ama_appeal` and `legacy_appeal`, along with other supplemental methods.

To [enable joining with polymorphic associations](https://stackoverflow.com/questions/16123492/eager-load-polymorphic):
- add to the ActiveRecord a scoped `belongs_to` that filters by polymorphic type
- add an instance method with the same name to return nil unless the association is the correct type

This address errors like `ActiveRecord::EagerLoadPolymorphicError: Cannot eagerly load the polymorphic association :appeal`.

(I attempted using Rails [refinements](https://docs.ruby-lang.org/en/2.4.0/syntax/refinements_rdoc.html) or [extend](https://longliveruby.com/articles/ruby-mixins-prepend-extend-include), but ended up with this PR that I think is more convenient to use.)

### Description
To enable efficient queries through a polymorphic association, this PR adds a concern that dynamically adds methods for polymorphic associations to decision reviews (appeals, legacy appeals, supplemental claims, higher level reviews) for `DecisionDocument` and `DecisionIssue`.

See code comments in `BelongsToPolymorphicAppealConcern`.

### Acceptance Criteria
- [ ] Code compiles correctly and all tests pass

### Testing Plan
See RSpec.

### CODE REVIEW

File: app/models/concerns/belongs_to_polymorphic_appeal_concern.rb [hunk 1]
```diff
@@ -0,0 +1,89 @@
+# frozen_string_literal: true
+
+require "helpers/association_wrapper.rb"
+
+##
+# This concern dynamically defines the following when `belongs_to_polymorphic_appeal :appeal` is called:
```
Comment (yoomlam) [inline app/models/concerns/belongs_to_polymorphic_appeal_concern.rb:6]:
```suggestion
# This concern dynamically defines the following when `belongs_to_polymorphic_appeal :appeal` is called,
# where `base_table_name` is the table name of the model that is including this concern:

```

File: app/models/concerns/belongs_to_polymorphic_appeal_concern.rb [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+# This concern dynamically defines the following when `belongs_to_polymorphic_appeal :appeal` is called,
+# where `base_table_name` is the table name of the model that is including this concern:
+
+#   belongs_to :appeal, polymorphic: true
+#
+#   scope :ama, -> { where(type_column => "Appeal") }
+#   scope :legacy, -> { where(type_column => "LegacyAppeal") }
+#   belongs_to :ama_appeal,
+#      -> { includes(base_table_name).where(base_table_name => {'appeal_type' => "Appeal"}) },
+#      class_name: "Appeal", foreign_key: 'appeal_id', optional: true
```
Comment (yoomlam) [inline app/models/concerns/belongs_to_polymorphic_appeal_concern.rb:15]: Note the `base_table_name` is plural. If the association from the `Appeal` to the active model class is `has_one` (instead of `has_many`), such as `appeal.special_issue_list`, then we'd need to singularize the `base_table_name` to match the `special_issue_list` method name created by `has_one`.

**Comment by yoomlam** [comment]
This PR address appeals being the polymorphic type. There will be follow-on PR(s) for the hearings polymorphic type.

**Review by lowellrex — APPROVED** [review]
LGTM!
